### ISSUE TITLE
bug(walk) Routes without a path are not provided.

### ISSUE THREAD

**Issue by themihai** [gorilla/mux#166]
If you register a route without a Path it is not provided by Router.Walk

**Comment by elithrar** [comment]
@themihai Can you provide an example?

**Comment by themihai** [comment]
```
package main

import (
	"github.com/gorilla/mux"

)

func main() {

	router := mux.NewRouter()
	router.Host("example.com")
	fn := func(route *mux.Route, router *mux.Router, ancestors []*mux.Route) error {
		panic("this should panic but it doesn't")
		return nil
	}
	if err := router.Walk(fn); err != nil {
		panic(err)
	}
}
```

**Comment by elithrar** [comment]
That's correct: the `walkFn` is only called if there are routes to walk:

https://github.com/gorilla/mux/blob/master/mux.go#L284

```go
nc (r *Router) walk(walkFn WalkFunc, ancestors []*Route) error {
	for _, t := range r.routes {
		if t.regexp == nil || t.regexp.path == nil || t.regexp.path.template == "" {
			continue
		}

		err := walkFn(t, r, ancestors)
		if err == SkipRouter {
			continue
...
```

**Comment by themihai** [comment]
But router.Host("example.com") creates a new route, isn't it? I mean it's a bug, right? I think the same applies to router.Methods(). I was using Walk to debug some strange behaviour of the Host() method thus the reason of the report.

**Comment by elithrar** [comment]
@themihai The docs are misleading here: what do you expect should be returned if you have a `*Router` with just a `Host` but no actual routes (paths)?

**Comment by themihai** [comment]
@elithrar  One Router.Host() route could be just one of the routes not the only one. For example I would like all the requests to xapi.domain.net to return HTTP 401 (Gone) to signal that it is no longer available so I setup a Router.Host(xapi.domain.net).Handler(GoneHandler). The other subdomains may have different routes with paths etc.
  On `Walk` you could use `GetHostTemplate` to get the host template. As this function already exists I assume it has some use-cases. 
  Design-wise it would make the API more orthogonal. 

Tbh I don't really use/need this but I reported it because it doesn't match the documentation so I thought there could be other bugs. The documentation states `Walk walks the router and *all* its sub-routers, calling walkFn for *each* route in the tree. ` which is clearly not the case.

**Comment by themihai** [comment]
@elithrar I've found that even routes with paths are not provided, though I can't figure out why (some are provided and some are not provided) so AFAIC this is beyond undocumented.

**Comment by elithrar** [comment]
I haven't had time to debug this: what is the length of r.routes inside
r.walk under different conditions?

e.g. adding just a host, adding actual routes, etc. If you can also show
your router code alongside it will help.

**Comment by themihai** [comment]
The routes configuration looks pretty much like this. It seems that only the routes that have both Host and Path are returned

```
  {
    "Routes": [  {
      "Method": "OPTIONS"
   },
    {
      "Method": "POST",
      "Href": "/discover"
    },
    {
      "Method": "GET",
      "Href": "/{userId}/{sessionStatus}/{sessionId}",
        "Host": "session{_:(^$|^\\.sandbox$)}.example.com"
    },
    {
      "Method": "POST",
      "Href": "/authUserId"
    },
    {
      "Method": "POST",
      "Href": "/registerUserId"
    },
    {
      "Method": "POST",
      "Href": "/forgotPassword"
    },
    {
      "Method": "POST",
      "Href": "/resetPassword"
    },
    {
      "Method": "POST",
      "Href": "/changePassword",
      "Host": "signin.api{_:(^$|^\\.sandbox$)}.example.co.uk"
    },

    {
      "Method": "GET",
      "Href": "/checkUsernameAvailability"
    },
    {
      "Method": "GET",
      "Href": "/disableSession"
    },
    {
      "Method": "GET",
      "Href": "/getActiveSessions"
    },
    {
      "Method": "GET",
      "Href": "/userIdByEmail"
     },
    {
      "Method": "GET",
      "Href": "/emailByUserId"
    },
    {
      "Method": "POST",
      "Href": "/changeEmail"
  
    },
    {
      "Method": "POST",
      "Href": "/changeEmailUnconfirmed"
  
    },
    {
      "Method": "GET",
      "Href": "/confirmEmail"
    },
    {
      "Method": "POST",
      "Href": "/resendConfirmEmail"
    },
    {
      "Method": "GET",
      "Href": "/disableUserId"
    },
   {
      "Method": "GET",
      "Href": "/reinstateUserId"
    },
   {
      "Method": "GET",
      "Href": "/userInfo"
    },
    {
      "Method": "POST",
      "Href": "/identity/"
    },
    {
      "Method": "PUT",
      "Href": "/identity/{sub}"
    },
    {
      "Method": "PUT",
      "Href": "/identity/{sub}/expiration"
    },
  
    {
      "Method": "POST",
      "Href": "/token"
    },
    {
      "Method": "PUT",
      "Href": "/token/revoke"
    },
    {
      "Method": "HEAD",
      "Href": "/identity/{key}/iss"
 
    },
    {
      "Method": "GET",
      "Href": "/user/info"

    },
    {
      "Method": "GET",
      "Href": "/.well-known/openid-configuration"

    },
    {
      "Method": "GET",
      "Href": "/lookup"

    }
  ]
}

```

// The routes are decoded from json and then loaded using something as below

```
        router :=  mux.NewRouter().StrictSlash(true)
	for _, rt := range routes {
		route := m.router.Methods(rt.Method)
		if rt.Host != "" {
			route.Host(host)
		}

		if rt.Href != "" {
			route.Path(href)
		}
	}

```
// After the routes are loaded a "debug" route is loaded at /muxdebug which uses Walk to retrieve the routes registered and return them (as json). It seems that only the routes that have both a Host and a Path are returned.  To make "Walk" useful I think all the routes should be returned regardless how many "parts" they have (i.e. host, path , method ). That you can debug what exactly the router is routing. I expected Walk to work somehow like the `path/filepath -> Walk` from the std library except that instead of files you would have routes.
```
[{"Host":"session{_:(^$|^\\.sandbox$)}.example.com","Path":"/{userId}/{sessionStatus}/{sessionId}","Ancestors":null},{"Host":"signin.api{_:(^$|^\\.sandbox$)}.example.co.uk","Path":"/changePassword","Ancestors":null}]
```

**Comment by nmiyake** [comment]
@themihai @elithrar I believe this is resolved by #261 and can be closed.
